lib/sensors/src/tof.rs

```diff
@@ -0,0 +1,386 @@
+use defmt;
+use embassy_time::{Duration, Timer};
+use hyped_io::i2c::{HypedI2c, I2cError};
+
+/// ToF implements the logic to read Time of Flight data from the VL6180V1 Time of Flight
+/// sensor using I2C peripheral provided by the Hyped I2c trait.
+///
+/// The majority of this implementation was done by implementing code examples from the Application Sheet (see below)
+/// into Rust code; this implementation should allow us to start single-shot and continuous measurements and read their results.
+/// Switching between single-shot and continuous measurements is outlined at the end of page 22 of the Application Sheet. The sensor's
+/// task in /tasks/tof.rs is a test task that reads range via single-shot measurement.
+///
+/// Data Sheet: https://www.st.com/en/imaging-and-photonics-solutions/vl6180.html#overview
+///
+/// Application Sheet: https://www.st.com/resource/en/application_note/an4545-vl6180x-basic-ranging-application-note-stmicroelectronics.pdf
+pub struct TimeOfFlight<'a, T: HypedI2c + 'a> {
+    i2c: &'a mut T,
+    device_address: u8,
+}
+
+impl<'a, T: HypedI2c> TimeOfFlight<'a, T> {
+    // Create a new instance of time of flight sensor, configure
+    pub fn new(i2c: &'a mut T, device_address: ToFAddresses) -> Result<Self, ToFError> {
+        let device_address = device_address as u8;
+
+        // Check that the sensor has powered up
+        // Note: the sensor will either need to be power cycled or the SYS_FRESH_OUT_RESET register will need to be written to 0x01 if it has already been configured
+        let mut boot_status = i2c
+            .read_byte_16(device_address, SYSTEM_FRESH_OUT_OF_RESET)
+            .unwrap_or_default();
+        while boot_status != 1 {
+            boot_status = i2c
+                .read_byte_16(device_address, SYSTEM_FRESH_OUT_OF_RESET)
+                .unwrap_or_default();
+        }
+        defmt::info!("ToF sensor booted");
```

Comment:
Thinking about this again: To take advantage of the `SYSTEM_FRESH_OUT_OF_RESET` register, you could just check if it is configured and if it is, skip initialising all of the registers (just extract that logic into an `init_sensor` (or similar) function